src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/service/VisitService.kt

```diff
@@ -308,6 +359,128 @@ class VisitService(
     }
   }
 
+  private fun getOrCreateVisitSlot(
+    location: AgencyLocation,
+    startDateTime: LocalDateTime,
+    endDateTime: LocalDateTime,
+  ): AgencyVisitSlot {
+    val agencyVisitTime =
+      getOrCreateVisitTime(startDateTime = startDateTime, endDateTime = endDateTime, location = location)
+    val vsipVisitRoom = getOrCreateVsipVisitRoom(location = location)
+    return visitSlotRepository.findByAgencyInternalLocation_DescriptionAndAgencyVisitTime_StartTimeAndWeekDay(
+      roomDescription = vsipVisitRoom.description,
+      startTime = agencyVisitTime.startTime,
+      weekDay = agencyVisitTime.agencyVisitTimesId.weekDay
+    ) ?: createVisitSlot(
+      visitTime = agencyVisitTime, vsipRoom = vsipVisitRoom
+    )
+  }
+
+  private fun getOrCreateVisitTime(
+    location: AgencyLocation,
+    startDateTime: LocalDateTime,
+    endDateTime: LocalDateTime,
+  ): AgencyVisitTime {
+    val weekDay = getOrCreateVisitDay(startDateTime = startDateTime, location = location)
+    return visitTimeRepository.findByStartTimeAndAgencyVisitTimesId_WeekDayAndAgencyVisitTimesId_Location(
+      startTime = startDateTime.toLocalTime(),
```

Comment:
so I guess we don't care if we get a slot where the end time doesn't match (which feels OK to me)